### DISCUSSION TITLE
Linker Error

### DISCUSSION THREAD

**Discussion by Ajayanand2023** [fengjiongmax/fltgst#9] [Q&A]
Can you help me solve this? My Android sdk version is 35. 

Starting process 'command '/home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja''. Working directory: /home/ajayanand/Ajay/Projects/Gstreamer/fltgst/app/native_binding/android
Command: /home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja -C /home/ajayanand/Ajay/Projects/Gstreamer/fltgst/app/native_binding/android/.cxx/Debug/53c3d52e/arm64-v8a native_binding
[        ] Successfully started process 'command '/home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja''
[        ] C/C++: ninja: Entering directory `/home/ajayanand/Ajay/Projects/Gstreamer/fltgst/app/native_binding/android/.cxx/Debug/53c3d52e/arm64-v8a'
[        ] C/C++: : && /home/ajayanand/Android/Sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=aarch64-none-linux-android21
--sysroot=/home/ajayanand/Android/Sdk/ndk/25.1.8937393/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables
-fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security  -fno-limit-debug-info  -static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment
-Wl,--fatal-warnings -Wl,--gc-sections -Wl,--no-undefined -Qunused-arguments -shared -Wl,-soname,libnative_binding.so -o
/home/ajayanand/Ajay/Projects/Gstreamer/fltgst/app/fltgst/build/native_binding/intermediates/cxx/Debug/53c3d52e/obj/arm64-v8a/libnative_binding.so
CMakeFiles/native_binding.dir/native_binding.c.o CMakeFiles/native_binding.dir/gst_plugin_init_android.c.o
-L/home/ajayanand/Ajay/Projects/Gstreamer/fltgst/app/native_binding/src/../third-party/gst-android/arm64/lib
-L/home/ajayanand/Ajay/Projects/Gstreamer/fltgst/app/native_binding/src/../third-party/gst-android/arm64/lib/gstreamer-1.0 -L/snap/flutter/current/usr/lib/x86_64-linux-gnu
-L/usr/lib/x86_64-linux-gnu  -lgstreamer-1.0  -lgobject-2.0  -lglib-2.0  -lgstcoreelements  -lgstcoretracers  -lgstadder  -lgstapp  -lgstaudioconvert  -lgstaudiorate  -lgstaudiotestsrc
-lgstgio  -lgstautodetect  -lgstopensles  -lintl  -lffi  -liconv  -lgmodule-2.0  -lpcre2-8  -lgstbase-1.0  -lgstaudio-1.0  -lorc-0.4  -lgstapp-1.0  -lgio-2.0  -llog  -lz  -lOpenSLES
-latomic -lm && :
[   +2 ms] C/C++: ld: error: --fix-cortex-a53-843419 is only supported on AArch64 targets
[        ] C/C++: clang: error: linker command failed with exit code 1 (use -v to see invocation)
[        ] C/C++: ninja: build stopped: subcommand failed.
[        ] Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
[        ] You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
[        ] For more on this, please refer to https://docs.gradle.org/8.3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
[        ] 50 actionable tasks: 5 executed, 45 up-to-date
[   +3 ms] FAILURE: Build failed with an exception.

**Comment by fengjiongmax** [comment]
The `--fix-cortex-a53-843419` was introduced by the toolchain, so maybe try to rebuild the project in a freshly set environment.
Reference this issue: https://github.com/fengjiongmax/fltgst/issues/1

**Reply by Ajayanand2023** [reply]
[        ] C/C++: Build native_binding
[        ] C/C++: /home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja \
[        ]   -C \
[        ]   /home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/android/.cxx/Debug/5m1bvh6p/arm64-v8a \
[        ]   native_binding
[        ] Starting process 'command '/home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja''. Working directory: /home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/android Command:
/home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja -C /home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/android/.cxx/Debug/5m1bvh6p/arm64-v8a native_binding
[   +1 ms] Successfully started process 'command '/home/ajayanand/Android/Sdk/cmake/3.22.1/bin/ninja''
[        ] C/C++: ninja: Entering directory `/home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/android/.cxx/Debug/5m1bvh6p/arm64-v8a'
[        ] C/C++: ld: error: undefined symbol: gst_init
[        ] C/C++: ld: error: undefined symbol: gst_element_factory_make
[        ] C/C++: ld: error: undefined symbol: gst_pipeline_new
[        ] C/C++: ld: error: undefined symbol: g_main_loop_new
[        ] C/C++: ld: error: undefined symbol: g_printerr
[        ] C/C++: ld: error: undefined symbol: gst_object_unref
[        ] C/C++: ld: error: undefined symbol: gst_bin_get_type
[        ] C/C++: ld: error: undefined symbol: g_type_check_instance_cast
[        ] C/C++: ld: error: undefined symbol: gst_bin_add_many
[        ] C/C++: ld: error: undefined symbol: gst_element_link_many
[        ] C/C++: ld: error: undefined symbol: gst_element_set_state
[        ] C/C++: ld: error: undefined symbol: g_main_loop_unref
[        ] C/C++: ld: error: undefined symbol: gst_plugin_register_static
[        ] C/C++: ld: error: undefined symbol: g_once_init_enter
[        ] C/C++: ld: error: undefined symbol: g_once_init_leave
[        ] C/C++: ld: error: undefined symbol: g_intern_static_string
[        ] C/C++: ld: error: undefined symbol: g_type_register_static_simple
[        ] C/C++: ld: error: undefined symbol: _gst_debug_category_new
[        ] C/C++: ld: error: undefined symbol: gst_element_register
[        ] C/C++: ld: error: undefined symbol: g_type_class_peek_parent
[        ] C/C++: ld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
[        ] C/C++: clang: error: linker command failed with exit code 1 (use -v to see invocation)
[        ] :native_binding:buildCMakeDebug[arm64-v8a] (Thread[Execution worker,5,main]) completed. Took 0.359 secs.
[        ] FAILURE: Build failed with an exception.
[        ] * What went wrong:
[        ] Execution failed for task ':native_binding:buildCMakeDebug[arm64-v8a]'.
[        ] > com.android.ide.common.process.ProcessException: ninja: Entering directory `/home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/android/.cxx/Debug/5m1bvh6p/arm64-v8a'
[        ]   [1/3] Building C object CMakeFiles/native_binding.dir/gst_plugin_init_android.c.o
[        ]   [2/3] Building C object CMakeFiles/native_binding.dir/native_binding.c.o
[        ]   [3/3] Linking C shared library /home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/fltgst/build/native_binding/intermediates/cxx/Debug/5m1bvh6p/obj/arm64-v8a/libnative_binding.so
[        ]   FAILED: /home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/fltgst/build/native_binding/intermediates/cxx/Debug/5m1bvh6p/obj/arm64-v8a/libnative_binding.so 
[        ]   : && /home/ajayanand/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=aarch64-none-linux-android21
--sysroot=/home/ajayanand/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fPIC -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong
-no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fexceptions -g  -fno-limit-debug-info  -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--fatal-warnings -Qunused-arguments
-Wl,--no-undefined -shared -Wl,-soname,libnative_binding.so -o
/home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/fltgst/build/native_binding/intermediates/cxx/Debug/5m1bvh6p/obj/arm64-v8a/libnative_binding.so CMakeFiles/native_binding.dir/native_binding.c.o
CMakeFiles/native_binding.dir/gst_plugin_init_android.c.o -L/home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/src/../third-party/gst-android/arm64/lib
-L/home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/src/../third-party/gst-android/arm64/lib/gstreamer-1.0 -lgstcoreelements  -lgstcoretracers  -lgstadder  -lgstapp
-lgstaudioconvert  -lgstaudiorate  -lgstaudiotestsrc  -lgstgio  -lgstautodetect  -lgstopensles  -lintl  -lffi  -liconv  -lgmodule-2.0  -lpcre2-8  -lgstbase-1.0  -lgstaudio-1.0  -lorc-0.4  -lgstapp-1.0
-lgio-2.0  -llog  -lz  -lOpenSLES   -latomic -lm && :
[   +1 ms]   ld: error: undefined symbol: gst_init
[        ]   >>> referenced by native_binding.c:26 (/home/ajayanand/Ajay/Git_project/flt_gst_check/fltgst/app/native_binding/src/native_binding.c:26)
[        ]   >>>               CMakeFiles/native_binding.dir/native_binding.c.o:(init)
[        ]   ld: error: undefined symbol: gst_element_factory_make


I have cloned the repository from git and tried building it with both cmake 3.18 as well as cmake 3.22.1. Also with gstreamer from old and new stable release. I get the above errors in all the cases. Can you help me solve the above issues.

**Reply by fengjiongmax** [reply]
You are getting different errors, which is a good sign.
This time the linker can not find the Gstreamer for Android libraries, make sure you have downloaded the Gstreamer for Android binaries and put them in the structure like the following:
```
native_binding
- third-party
   - gst-android
     - armv7
     - arm64
     - x86
     - x86_64
```

**Reply by Ajayanand2023** [reply]
# This CMakeLists.txt solved the above error. Thanks for your assistance.
cmake_minimum_required(VERSION 3.10)
project(native_binding_library VERSION 0.0.1 LANGUAGES C)
find_package(PkgConfig REQUIRED)

pkg_search_module(GST REQUIRED gstreamer-1.0)

IF(WIN32)
  find_program(CMAKE_PKGCONFIG_EXECUTABLE pkg-config)
  IF(CMAKE_PKGCONFIG_EXECUTABLE)
    EXEC_PROGRAM(${CMAKE_PKGCONFIG_EXECUTABLE}
    ARGS " --libs --msvc-syntax gstreamer-1.0"
    OUTPUT_VARIABLE GST_LDFLAGS)
    string(REPLACE " " ";" GST_LDFLAGS ${GST_LDFLAGS})
    message("GST_LDFLAGS: ${GST_LDFLAGS}")
  ENDIF()
ENDIF()

IF(ANDROID OR __ANDROID__)
  SET(GST_FOLDER ${CMAKE_CURRENT_SOURCE_DIR}/../third-party/gst-android)
  SET(ABI_SUFFIX ${ANDROID_ABI})

  IF(${ANDROID_ABI} STREQUAL "armeabi-v7a")
    SET(ABI_SUFFIX armv7)
  ELSEIF(${ANDROID_ABI} STREQUAL "arm64-v8a")
    SET(ABI_SUFFIX arm64)
  ELSEIF(${ANDROID_ABI} STREQUAL "x86")
    # skipped
  ELSEIF(${ANDROID_ABI} STREQUAL "x86_64")
    # skipped
  ENDIF()

  SET(GST_ROOT ${GST_FOLDER}/${ABI_SUFFIX})

  include_directories(
    ${GST_ROOT}/include/gstreamer-1.0
    ${GST_ROOT}/include/glib-2.0
    ${GST_ROOT}/lib/glib-2.0/include
  )

  link_directories(
    ${GST_ROOT}/lib
    ${GST_ROOT}/lib/gstreamer-1.0
  )

  SET(PLUGINS_DECLARATION)
  SET(PLUGINS_REGISTRATION)

  LIST(APPEND GST_PLUGINS coreelements coretracers adder app audioconvert audiorate audiotestsrc gio autodetect opensles)
  foreach(GST_P ${GST_PLUGINS})
    LIST(APPEND LINK_LIBS "gst${GST_P}")
    LIST(APPEND PLUGINS_DECLARATION "\nGST_PLUGIN_STATIC_DECLARE(${GST_P})")
    LIST(APPEND PLUGINS_REGISTRATION "\nGST_PLUGIN_STATIC_REGISTER(${GST_P})")
  endforeach()

  configure_file(gst_android.c.in ${CMAKE_CURRENT_SOURCE_DIR}/gst_plugin_init_android.c)

  LIST(APPEND APPENDED_SOURCE gst_plugin_init_android.c)

  LIST(APPEND LINK_LIBS intl ffi iconv gmodule-2.0 pcre2-8 gstbase-1.0 gstaudio-1.0 orc-0.4 gstapp-1.0 gio-2.0 log z OpenSLES)

ELSE()
  include_directories(${GST_INCLUDE_DIRS})
ENDIF()

add_library(native_binding SHARED
  "native_binding.c"
  ${APPENDED_SOURCE}
)

IF(WIN32)
  target_link_options(native_binding PRIVATE ${GST_LDFLAGS})
ELSE()
  target_link_libraries(native_binding PRIVATE ${GST_LDFLAGS})
  target_link_libraries(native_binding PRIVATE ${GST_LDFLAGS} ${LINK_LIBS})
ENDIF()

set_target_properties(native_binding PROPERTIES
  PUBLIC_HEADER native_binding.h
  OUTPUT_NAME "native_binding"
)
target_compile_definitions(native_binding PUBLIC DART_SHARED_LIB)

**Comment by Ajayanand2023** [comment]
Can you make a explanatory video on Video display

**Reply by fengjiongmax** [reply]
Yes but not soon, because video display doesn't work on Windows and Linux yet, but you can check the #2 issue for an overview of the cross-language and platform integration.

**Reply by Ajayanand2023** [reply]
Hi, Thanks as always. I tried modifying the native_binding.c, so that without constructing the pipeline manually, I used gst_parse_launch to handle the pipeline. It worked well with videotestsrc, but when i tried to play a uri or rtsp stream, I don't get any error, but the video is not played. I also included the necessary libraries:
1). libgstrtsp - rtspsrc
2). libgstrtp - rtph265depay
3). libgstvideoparsersbad - h265parse
4). libgstplayback - decodebin
5). libgstvideoconvertscale - videoconvert
6). libgstautodetect - autovideosink

#include "native_binding.h"
#include <gst/gst.h>
#include <stdio.h>
#include <gst/video/video.h>

#if ANDROID
#include <jni.h>
#include <android/log.h>
#include <android/native_window.h>
#include <android/native_window_jni.h>
#endif

#ifdef IOS
#include "gst_ios_init.h"
#endif

typedef struct _FltGstData
{
    GstElement *pipeline;
    GMainLoop *mainloop;

    GstElement *overlay;

#if ANDROID
    ANativeWindow *native_window;
#endif
} FltGstData;

FltGstData *data;

FFI_PLUGIN_EXPORT void init(void)
{
#ifdef IOS
    gst_ios_init();
#endif
    // init
    gst_init(NULL, NULL);
    data = g_new0(FltGstData,1);
}

FFI_PLUGIN_EXPORT void setup_pipeline(void)
{
    // setup pipeline
    GError *error = NULL;
    char *rtsp_url = "rtsp://192.168.0.139/live/main";
    char pipeline_description[512];
    snprintf(pipeline_description, sizeof(pipeline_description),
             "rtspsrc location=%s latency=100 ! rtph265depay ! h265parse ! avdec_h265 ! videoconvert ! autovideosink",
             rtsp_url);

    // Parse and create the pipeline
    data->pipeline = gst_parse_launch(pipeline_description, &error);
    data->mainloop = g_main_loop_new(NULL, FALSE);

    if (!data->pipeline) {
        g_printerr("Error creating pipeline: %s\n", error->message);
        g_clear_error(&error);
        return;
    }
    gst_element_set_state (data->pipeline, GST_STATE_READY);

    data->overlay = gst_bin_get_by_interface(GST_BIN(data->pipeline),GST_TYPE_VIDEO_OVERLAY);

}


FFI_PLUGIN_EXPORT void start_pipeline(void)
{
    // start pipeline
    GstStateChangeReturn ret;
    ret = gst_element_set_state(data->pipeline, GST_STATE_PLAYING);
    if (ret == GST_STATE_CHANGE_FAILURE)
    {
        g_printerr("Elements could not be created.\n");
        gst_object_unref(data->pipeline);
        return;
    }
}


void set_native_window(gpointer native_window_ptr)
{
    GstVideoOverlay  *overlay = GST_VIDEO_OVERLAY(data->overlay);
    if (native_window_ptr && data->pipeline)
    {
        gst_video_overlay_set_window_handle(overlay,(guintptr)native_window_ptr);
        
        GstState cur_state;
        gst_element_get_state(GST_ELEMENT(data->pipeline),&cur_state,NULL,0);
        g_print("cur_state: %d",cur_state);
    }
}


#if ANDROID

JNIEXPORT void JNICALL
Java_com_example_native_1view_1example_NativeView_nativeSurfaceFianlize(JNIEnv *env, jobject thiz)
{
    // TODO: implement nativeSurfaceFianlize()
    GST_DEBUG("Releasing Native Window %p", data->native_window);
    if (data->pipeline)
    {
        gst_video_overlay_set_window_handle(GST_VIDEO_OVERLAY(data->overlay),
                                            (guintptr)NULL);
        gst_element_set_state(data->pipeline, GST_STATE_READY);
    }

    ANativeWindow_release(data->native_window);
    data->native_window = NULL;
}

JNIEXPORT void JNICALL
Java_com_example_native_1view_1example_NativeView_nativeSurfaceInit(JNIEnv *env, jobject thiz,
                                                                    jobject surface)
{
    // TODO: implement nativeSurfaceInit()
    ANativeWindow *new_native_window = ANativeWindow_fromSurface(env, surface);
    GST_DEBUG("Received surface %p (native window %p)", surface,
              new_native_window);

    if (data->native_window)
    {
        ANativeWindow_release(data->native_window);
        if (data->native_window == new_native_window)
        {
            GST_DEBUG("New native window is the same as the previous one %p",
                      data->native_window);
            if (data->pipeline)
            {
                gst_video_overlay_expose(GST_VIDEO_OVERLAY(data->pipeline));
                gst_video_overlay_expose(GST_VIDEO_OVERLAY(data->pipeline));
            }
            return;
        }
        else
        {
            GST_DEBUG("Released previous native window %p", data->native_window);
            //            data->initialized = FALSE;
        }
    }
    data->native_window = new_native_window;

    GstVideoOverlay  *overlay = GST_VIDEO_OVERLAY(data->overlay);
    if (data->native_window && data->pipeline)
    {
        gst_video_overlay_set_window_handle(overlay,(guintptr)data->native_window);
    }

    GstState cur_state;
    gst_element_get_state(GST_ELEMENT(data->pipeline),&cur_state,NULL,0);

}

#endif

FFI_PLUGIN_EXPORT void free_resource(void)
{
    // free resources
    gst_element_set_state(data->pipeline, GST_STATE_NULL);
    gst_object_unref(data->pipeline);
    g_free(data);
}

// A very short-lived native function.
//
// For very short-lived functions, it is fine to call them on the main isolate.
// They will block the Dart execution while running the native function, so
// only do this for native functions which are guaranteed to be short-lived.
FFI_PLUGIN_EXPORT intptr_t sum(intptr_t a, intptr_t b) { return a + b; }

// A longer-lived native function, which occupies the thread calling it.
//
// Do not call these kind of native functions in the main isolate. They will
// block Dart execution. This will cause dropped frames in Flutter applications.
// Instead, call these native functions on a separate isolate.
FFI_PLUGIN_EXPORT intptr_t sum_long_running(intptr_t a, intptr_t b)
{
    // Simulate work.
#if _WIN32
    Sleep(5000);
#else
    usleep(5000 * 1000);
#endif
    return a + b;
}

**Reply by fengjiongmax** [reply]
I noticed you didn't mention any libraries added for the `avdec_h265` element, probably this is the reason.
The reason I create the elements separately and then assemble the pipeline is, that I can check which element was not created, and then I can fix it, considering your pipeline is not complicated, I suggest you convert your pipeline assemble method, **or you can try to create the elements separately and check if they are null when debugging**.

**Comment by Ajayanand2023** [comment]
I even tried the android tutorial 3 from gstreamer to play audio from uridecodebin and added the necessary plugins including soup for http request, but I couldn't hear the audio. The same works fine in Linux application. Can you help me resolve this for Android flutter.

#include "native_binding.h"
#include <gst/gst.h>

#ifdef IOS
    #include "gst_ios_init.h"
#endif

typedef struct _CustomData {
  GstElement *pipeline;
  GstElement *source;
  GstElement *convert;
  GstElement *resample;
  GstElement *sink;

  GMainLoop *mainloop;
} CustomData;

CustomData *data;

static void pad_added_handler (GstElement *src, GstPad *pad, CustomData *data);

FFI_PLUGIN_EXPORT void init(void)
{
#ifdef IOS
    gst_ios_init();
#endif
    gst_init(NULL, NULL);
    data = (CustomData *)malloc(sizeof(CustomData));
}

FFI_PLUGIN_EXPORT void setup_pipeline(void)
{
    // setup pipeline
    data->source = gst_element_factory_make ("uridecodebin", "source");
    data->convert = gst_element_factory_make ("audioconvert", "convert");
    data->resample = gst_element_factory_make ("audioresample", "resample");
    data->sink = gst_element_factory_make ("autoaudiosink", "sink");
    data->pipeline = gst_pipeline_new ("test-pipeline");

    data->mainloop = g_main_loop_new(NULL, FALSE);

    if (!data->pipeline || !data->source || !data->convert || !data->resample || !data->sink) {
        g_printerr ("Not all elements could be created.\n");
        return;
    }

    gst_bin_add_many (GST_BIN (data->pipeline), data->source, data->convert, data->resample, data->sink, NULL);
    if (!gst_element_link_many (data->convert, data->resample, data->sink, NULL)) {
        g_printerr ("Elements could not be linked.\n");
        gst_object_unref (data->pipeline);
        return;
    }

    g_object_set (data->source, "uri", "https://gstreamer.freedesktop.org/data/media/sintel_trailer-480p.webm", NULL);

    g_signal_connect (data->source, "pad-added", G_CALLBACK (pad_added_handler), data);
}

FFI_PLUGIN_EXPORT void start_pipeline(void)
{
    GstStateChangeReturn ret;
    ret = gst_element_set_state(data->pipeline, GST_STATE_PLAYING);
    if (ret == GST_STATE_CHANGE_FAILURE)
    {
        g_printerr("Elements could not be created.\n");
        gst_object_unref(data->pipeline);
        return;
    }
}

FFI_PLUGIN_EXPORT void free_resource(void)
{
    g_main_loop_unref(data->mainloop);
    gst_element_set_state(data->pipeline, GST_STATE_NULL);
    gst_object_unref(data->pipeline);
}

static void pad_added_handler (GstElement *src, GstPad *new_pad, CustomData *data) {
  GstPad *sink_pad = gst_element_get_static_pad (data->convert, "sink");
  GstPadLinkReturn ret;
  GstCaps *new_pad_caps = NULL;
  GstStructure *new_pad_struct = NULL;
  const gchar *new_pad_type = NULL;

  g_print ("Received new pad '%s' from '%s':\n", GST_PAD_NAME (new_pad), GST_ELEMENT_NAME (src));

  if (gst_pad_is_linked (sink_pad)) {
    g_print ("We are already linked. Ignoring.\n");
    goto exit;
  }

  new_pad_caps = gst_pad_get_current_caps (new_pad);
  new_pad_struct = gst_caps_get_structure (new_pad_caps, 0);
  new_pad_type = gst_structure_get_name (new_pad_struct);
  if (!g_str_has_prefix (new_pad_type, "audio/x-raw")) {
    g_print ("It has type '%s' which is not raw audio. Ignoring.\n", new_pad_type);
    goto exit;
  }

  ret = gst_pad_link (new_pad, sink_pad);
  if (GST_PAD_LINK_FAILED (ret)) {
    g_print ("Type is '%s' but link failed.\n", new_pad_type);
  } else {
    g_print ("Link succeeded (type '%s').\n", new_pad_type);
  }

exit:
  if (new_pad_caps != NULL)
    gst_caps_unref (new_pad_caps);

  /* Unreference the sink pad */
  gst_object_unref (sink_pad);
}

// A very short-lived native function.
//
// For very short-lived functions, it is fine to call them on the main isolate.
// They will block the Dart execution while running the native function, so
// only do this for native functions which are guaranteed to be short-lived.
FFI_PLUGIN_EXPORT intptr_t sum(intptr_t a, intptr_t b) { return a + b; }

// A longer-lived native function, which occupies the thread calling it.
//
// Do not call these kind of native functions in the main isolate. They will
// block Dart execution. This will cause dropped frames in Flutter applications.
// Instead, call these native functions on a separate isolate.
FFI_PLUGIN_EXPORT intptr_t sum_long_running(intptr_t a, intptr_t b)
{
    // Simulate work.
#if _WIN32
    Sleep(5000);
#else
    usleep(5000 * 1000);
#endif
    return a + b;
}

**Reply by fengjiongmax** [reply]
Does your app have the internet access permission?

**Comment by Ajayanand2023** [comment]
Yes, I gave the permission to manifest file.

**Comment by Ajayanand2023** [comment]
This is my Manifest file
```

<manifest xmlns:android="http://schemas.android.com/apk/res/android">
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE" />
    <application
        android:label="fltgst"
        android:name="${applicationName}"
        android:icon="@mipmap/ic_launcher">
        <activity
            android:name=".MainActivity"
            android:exported="true"
            android:launchMode="singleTop"
            android:theme="@style/LaunchTheme"
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|smallestScreenSize|locale|layoutDirection|fontScale|screenLayout|density|uiMode"
            android:hardwareAccelerated="true"
            android:windowSoftInputMode="adjustResize">
            <!-- Specifies an Android theme to apply to this Activity as soon as
                 the Android process has started. This theme is visible to the user
                 while the Flutter UI initializes. After that, this theme continues
                 to determine the Window background behind the Flutter UI. -->
            <meta-data
              android:name="io.flutter.embedding.android.NormalTheme"
              android:resource="@style/NormalTheme"
              />
            <intent-filter>
                <action android:name="android.intent.action.MAIN"/>
                <category android:name="android.intent.category.LAUNCHER"/>
            </intent-filter>
        </activity>
        <!-- Don't delete the meta-data below.
             This is used by the Flutter tool to generate GeneratedPluginRegistrant.java -->
        <meta-data
            android:name="flutterEmbedding"
            android:value="2" />
    </application>
</manifest>

```

**Reply by Ajayanand2023** [reply]
This is my CMakeLists.txt

```
cmake_minimum_required(VERSION 3.10)
project(native_binding_library VERSION 0.0.1 LANGUAGES C)
find_package(PkgConfig REQUIRED)

pkg_search_module(GST REQUIRED gstreamer-1.0)

IF(WIN32)
find_program(CMAKE_PKGCONFIG_EXECUTABLE pkg-config)
IF(CMAKE_PKGCONFIG_EXECUTABLE)
EXEC_PROGRAM(${CMAKE_PKGCONFIG_EXECUTABLE}
ARGS " --libs --msvc-syntax gstreamer-1.0"
OUTPUT_VARIABLE GST_LDFLAGS)
string(REPLACE " " ";" GST_LDFLAGS ${GST_LDFLAGS})
message("GST_LDFLAGS: ${GST_LDFLAGS}")
ENDIF()
ENDIF()

IF(ANDROID OR ANDROID)
SET(GST_FOLDER ${CMAKE_CURRENT_SOURCE_DIR}/../third-party/gst-android)
SET(ABI_SUFFIX ${ANDROID_ABI})

IF(${ANDROID_ABI} STREQUAL "armeabi-v7a")
SET(ABI_SUFFIX armv7)
ELSEIF(${ANDROID_ABI} STREQUAL "arm64-v8a")
SET(ABI_SUFFIX arm64)
ELSEIF(${ANDROID_ABI} STREQUAL "x86")
# skipped
ELSEIF(${ANDROID_ABI} STREQUAL "x86_64")
# skipped
ENDIF()

SET(GST_ROOT ${GST_FOLDER}/${ABI_SUFFIX})

include_directories(
${GST_ROOT}/include/gstreamer-1.0
${GST_ROOT}/include/glib-2.0
${GST_ROOT}/lib/glib-2.0/include
${GST_ROOT}/include/libsoup-2.4
)

link_directories(
${GST_ROOT}/lib
${GST_ROOT}/lib/gstreamer-1.0
)

SET(PLUGINS_DECLARATION)
SET(PLUGINS_REGISTRATION)

LIST(APPEND GST_PLUGINS coreelements coretracers adder app audioconvert audiorate audiotestsrc gio autodetect opensles soup playback audioresample)
foreach(GST_P ${GST_PLUGINS})
LIST(APPEND LINK_LIBS "gst${GST_P}")
LIST(APPEND PLUGINS_DECLARATION "\nGST_PLUGIN_STATIC_DECLARE(${GST_P})")
LIST(APPEND PLUGINS_REGISTRATION "\nGST_PLUGIN_STATIC_REGISTER(${GST_P})")
endforeach()
LIST(APPEND LINK_LIBS soup-2.4)

configure_file(gst_android.c.in ${CMAKE_CURRENT_SOURCE_DIR}/gst_plugin_init_android.c)

LIST(APPEND APPENDED_SOURCE gst_plugin_init_android.c)

LIST(APPEND LINK_LIBS intl ffi iconv gmodule-2.0 pcre2-8 gstbase-1.0 gstaudio-1.0 orc-0.4 gstapp-1.0 gio-2.0 log z OpenSLES gstavi gstisomp4 soup-2.4 psl gstsoup gsticydemux gstjpegformat gstriff-1.0 gsttag-1.0 ges-1.0 gstpbutils-1.0 gstencoding gstcodecalpha gstcamerabin gsthls gstreplaygain gstplayback gstcodecs-1.0 gstvideo-1.0 gstvpx gstpng gstopenh264 gstpnm gstopenjpeg gstjpeg gsttheora gstlibav gstandroidmedia gstrsvg)

ELSE()
include_directories(${GST_INCLUDE_DIRS})
ENDIF()

add_library(native_binding SHARED
"native_binding.c"
${APPENDED_SOURCE}
)

IF(WIN32)
target_link_options(native_binding PRIVATE ${GST_LDFLAGS})
ELSE()
target_link_libraries(native_binding PRIVATE ${GST_LDFLAGS})
target_link_libraries(native_binding PRIVATE ${GST_LDFLAGS} ${LINK_LIBS})
ENDIF()

set_target_properties(native_binding PROPERTIES
PUBLIC_HEADER native_binding.h
OUTPUT_NAME "native_binding"
)
target_compile_definitions(native_binding PUBLIC DART_SHARED_LIB)
```

**Comment by Ajayanand2023** [comment]
@fengjiongmax Hi, Have you tried playing audio using filesrc or uridecodebin. In both case, my android app is working but not playing audio. But my linux application is working fine. I also included plugins and libraries through gst-inspect-1.0 and grep to lib folders. Do you have any suggestions?

**Reply by fengjiongmax** [reply]
I haven't had time to try uridecodebin yet, my guess is, that the uridecodebin will create some elements that you haven't added the libraries for, you can start by checking what's inside that uridecodebin in your pipeline.

https://gstreamer.freedesktop.org/documentation/gstreamer/gstbin.html?gi-language=c#GstBin
